File: baekjoon/p12018/Main.java
package baekjoon.p12018;

import java.util.Arrays;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();

        P12018[] inputs = new P12018[n];
        for (int i = 0; i < inputs.length; i++) {
            int p = sc.nextInt();
            int l = sc.nextInt();

            int[] arr = new int[p];
            for (int j = 0; j < arr.length; j++) {
                arr[j] = sc.nextInt();
            }

            inputs[i] = new P12018(p, l, arr);
        }

        System.out.println(solve(n, m, inputs));
    }

    private static int solve(int n, int m, P12018[] inputs) {
        int[] min = new int[n];
        for (int i = 0; i < min.length; i++) {
            min[i] = inputs[i].getMin();
        }

        Arrays.sort(min);
        int sum = 0;
        for (int i = 0; i < min.length; i++) {
            sum += min[i];
            if (sum > m) {
                return i;
            }
        }

        return n;
    }

    private static class P12018 {

        private int p;
        private int l;
        private int[] arr;

        P12018(int p, int l, int[] arr) {
            this.p = p;
            this.l = l;
            this.arr = arr;
        }

        int getMin() {
            if (p < l) {
                return 1;
            } else {
                Arrays.sort(arr);
                return arr[p - l];
            }
        }

    }

}

File: baekjoon/p12018/Main.kt
package baekjoon.p12018

fun main() {
    val (n, m) = readLine()!!.split(" ").map { it.toInt() }
    val inputs = List(n) {
        val (p, l) = readLine()!!.trim().split(" ").map { it.toInt() }
        val list = readLine()!!.split(" ").map { it.toInt() }
        P12018(p, l, list)
    }

    println(solve(n, m, inputs))
}

private fun solve(n: Int, m: Int, inputs: List<P12018>): Int {
    val min = (0 until n).map { inputs[it].getMin() }.sorted()
    var sum = 0

    for (i in 0 until min.size) {
        sum += min[i]
        if (sum > m) {
            return i
        }
    }

    return n
}

private class P12018(val p: Int, val l: Int, val list: List<Int>) {
    fun getMin(): Int {
        return when (p < l) {
            true -> 1
            false -> list.sorted()[p - l]
        }
    }
}